File: neet_code_150_cpp/backtracking/subsets.cpp
#include <iostream>
#include <vector>

// AC
// ビット演算で部分集合を数え上げる
class Solution {
    public:
        std::vector<std::vector<int>> subsets(std::vector<int>& nums) {
            auto n = (int)nums.size();
            std::vector<std::vector<int>> result;

            for(int i = 0; i < (1 << n); ++i) {
                std::vector<int> sub_set;
                for(int j = 0; j < n; ++j) {
                    if(i & (1 << j)) {
                        sub_set.push_back(nums[j]);
                    }
                }
                result.push_back(sub_set);
            }

            return result;
        }
};

// 模範解答
// DFS
class SolutionAns {
    public:
        std::vector<std::vector<int>> subsets(std::vector<int>& nums) {
            std::vector<int> current;
            std::vector<std::vector<int>> result;
            dfs(nums, 0, current, result);

            return result;
        }

    private:
        void dfs(
                std::vector<int>& nums, 
                int start, 
                std::vector<int>& current, 
                std::vector<std::vector<int>>& result
                ) {
            result.push_back(current);

            for(int i = start; i < nums.size(); ++i) {
                current.push_back(nums[i]);
                dfs(nums, i + 1, current, result);
                current.pop_back();
            }
        }
};

void show_result(const std::vector<std::vector<int>>& result) {
    for(auto& v: result) {
        for(auto& w: v) {
            std::cout << w << " ";
        }
        std::cout << '\n';
    }
}

int main(void) {
    auto case_1 = std::vector{1, 2, 3};
    auto case_2 = std::vector{0};

    Solution s_1;

    auto res_1 = s_1.subsets(case_1);
    auto res_2 = s_1.subsets(case_2);
    show_result(res_1);
    show_result(res_2);

    SolutionAns s_ans;
    auto res_1_ans = s_ans.subsets(case_1);
    auto res_2_ans = s_ans.subsets(case_2);
    show_result(res_1_ans);
    show_result(res_2_ans);
}

File: neet_code_150/src/backtracking/subsets.rs
use std::iter::FromIterator;

// AC
struct Solution {}
impl Solution {
    // O(N2^N)
    pub fn subsets(nums: Vec<i32>) -> Vec<Vec<i32>> {
        let n = nums.len();
        let mut result = vec![];

        for i in 0..(1 << n) {
            let mut sub_set = vec![];
            for j in 0..n {
                if i & (1 << j) != 0 {
                    sub_set.push(nums[j]);
                }
            }
            result.push(sub_set.clone());
        }

        result
    }
}

// AC
// C++の模範解答
struct SolutionAnsCpp {}
impl SolutionAnsCpp {
    pub fn subsets(nums: Vec<i32>) -> Vec<Vec<i32>> {
        let mut current = vec![];
        let mut result = vec![];

        Self::dfs(&nums, 0, &mut current, &mut result);

        result
    }

    fn dfs(nums: &Vec<i32>, start: i32, current: &mut Vec<i32>, result: &mut Vec<Vec<i32>>) {
        result.push(current.clone());

        for i in start..(nums.len() as i32) {
            current.push(nums[i as usize]);
            Self::dfs(nums, i + 1, current, result);
            current.pop();
        }
    }
}

// 模範解答
// ビット演算とイテレータ
struct SolutionAns {}
impl SolutionAns {
    pub fn subsets(nums: Vec<i32>) -> Vec<Vec<i32>> {
        let n = nums.len();

        Vec::from_iter((0..1 << n).map(|bitmask| {
            Vec::from_iter((0..n).filter_map(|i| match (bitmask >> i) & 1 != 0 {
                true => Some(nums[i]),
                false => None,
            }))
        }))
    }
}

fn main() {
    let case_1 = vec![1, 2, 3];
    let case_2 = vec![0];

    println!("case_1: {:?}", Solution::subsets(case_1.clone()));
    println!("case_2: {:?}", Solution::subsets(case_2.clone()));

    println!("case_1: {:?}", SolutionAnsCpp::subsets(case_1.clone()));
    println!("case_2: {:?}", SolutionAnsCpp::subsets(case_2.clone()));

    println!("case_1: {:?}", SolutionAns::subsets(case_1.clone()));
    println!("case_2: {:?}", SolutionAns::subsets(case_2.clone()));
}
